### ISSUE TITLE
Unable to find SAS drives (works with older sedutil)

### ISSUE THREAD

**Issue by surendarchandra** [Drive-Trust-Alliance/sedutil#200]
I have a Hitachi HUH728080AL5205 drive and sedutil does not recognize it:

Scanning for Opal compliant disks
/dev/sda   No     

I have a older version of sedutil (and I have no idea where it came from because it has a file called linux/MsedDevLinuxSas.cpp) and that detects these drives!!
Scanning for Opal compliant disks
/dev/sda   E R5GBKH7V HUH728080AL5205  D907

**Comment by yaplej** [comment]
I have a bunch of multi-path SAS devices that sedutil does not recognize.

ST33000650SS

/sedutil-cli -vv --query /dev/disk/by-id/dm-uuid-mpath-[SASADDRESS]
DtaDevOS::init ERROR - unknown drive type
Invalid or unsupported disk /dev/disk/by-id/dm-uuid-mpath-[SASADDRESS]

I also tried using a single path.
./sedutil-cli -v --query /dev/sdbs
The Kernel flag libata.allow_tpm is not set correctly
Please see the readme note about setting the libata.allow_tpm
Invalid or unsupported disk /dev/sdbs

Was just trying to unlock them so they could be used like a normal device.

**Comment by r0m30** [comment]
I'm not that familar with SAS, the support was contributed and it is possible that support was broken along the way.  I'm not sure where to even look from the info here.

**Comment by pruiz** [comment]
Hello,

I just run into this same issue while working with some HGST (SAS) drives too, and here is a minimal patch fixing the issue (at least to the point of --scan / --query working ok). I will be doing some more testing this weekend and probably will be submitting a pull-request with the changes.

Meanwhile, just in case @surendarchandra wants to try by himself:

```
diff --git a/linux/DtaDevLinuxSata.cpp b/linux/DtaDevLinuxSata.cpp
index fa71bd3..461d1d8 100644
--- a/linux/DtaDevLinuxSata.cpp
+++ b/linux/DtaDevLinuxSata.cpp
@@ -246,7 +246,6 @@ void DtaDevLinuxSata::identify(OPAL_DiskInfo& disk_info)
         LOG(D4) << "sense after ";
         IFLOG(D4) DtaHexDump(sense, sizeof (sense));
         disk_info.devType = DEVICE_TYPE_OTHER;
-        identify_SAS(&disk_info);
         return;
     }
     //    LOG(D4) << "cdb after ";
@@ -271,11 +270,12 @@ void DtaDevLinuxSata::identify(OPAL_DiskInfo& disk_info)

     if (!(memcmp(nullz.data(), buffer, 512))) {
         disk_info.devType = DEVICE_TYPE_OTHER;
+        identify_SAS(&disk_info);
         return;
     }
```

**Comment by mfuerni** [comment]
Hello,

the fix proposed by @pruiz doesn't work with my Seagate Enterprise drives.
I tried it with some Seagate Enterprise Capacity 3.5 HDD 10TB, 512e, SED, SAS 12Gb/s disks.

```
# ./sedutil-cli -vvvvv --query /dev/sda
0000 12000000 2400                        ....$.
0000 00000000 00000000 00000000 00000000  ................
0010 0000                                 ..
Invalid or unsupported disk /dev/sda
```

**Comment by pruiz** [comment]
@mfuerni not sure, but maybe your drive is plugged into a raid controller (w/o passthru support?).

**Comment by mfuerni** [comment]
yes, I use a Lenovo/Broadcom 930-16i RAID controller in JBOD mode, but SeaTools on Windows can access SED features of the disk. So i think it supports pass-through.

**Comment by pruiz** [comment]
@mfuerni, It may support pass-thru (specially on windows), but not sure it will support it on linux. From the output of your sed-utils, this is going bananas much earlier than what my patch fixes.. it simply looks like it doesn't recognize sda as a SATA or SAS device..

**Comment by JonFraserNH** [comment]
I applied the patch and it fixed the problem I had with Seagate ST10000NM0226 Enterprise SAS drives not working.  The drives are TCG compliant, but not OPAL compliant, according to email I received from Seagate support.  The controller is an LSI SAS3008 PCI-Express Fusion-MPT SAS-3 (rev 02).

Info from the seagate SeaChest_security program:
sg1 - ST10000NM0226 - ZA25GGKH0000C828CCAK - SCSI
TCG Device Information:
TCG Type: TCG Enterprise
Data Access: Unlocked
Drive State: Use
Drive Ports:
        Firmware Download - Unlocked
        UDS - Locked
Security Operating Mode: 0 - No Security Active
Data Locking: Available
Supported TCG Features:
        TPER
        Locking
        Geometry
        Secure Messaging

Before the patch:
sedutil-cli --scan
Scanning for Opal compliant disks
/dev/sda    2  Samsung SSD 860 EVO 250GB                RVT01B6Q
/dev/sdb   No
/dev/sdc   No

After the patch:
Scanning for Opal compliant disks
/dev/sda    2  Samsung SSD 860 EVO 250GB                RVT01B6Q
/dev/sdb     E ST10000NM0226    KT02
/dev/sdc     E ST10000NM0226    KT02


Note that MBR is not supported.
sedutil-cli --setMBREnable off xxxxx /dev/sdb
MBR shadowing is optional in the Enterprise SSC and not supported

These are data, not boot, drives, so we don't want the MBR shadowing anyway.
I would prefer that --initialsetup  not manipulate the MBR settings. Perhaps an 'Enterprise' flag
would be useful.

FYI, initialSetup did produce a lot of errors, the same on all 4 drives I setup.
Here's the output in case it is useful.

sedutil-cli --initialsetup debug /dev/sdb
EraseMaster  password set
Maximum ranges supported 31
BandMaster0 password set
BandMaster1 password set
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
Session Authenticate failed (response = false)
takeOwnership complete
Locking range Read/Write set 0
Locking range configured 3
Initial setup of TPer complete on /dev/sdb
